### PR TITLE
Dashboard item description (1/2)

### PR DESCRIPTION
**Pull request by liyiy**
## Changes

(1/2) Updates for dashboard item editing on #3551  

Screen Shot 2021-05-06 at 3 06 17 PM

- Follow up PR (2/2) will include 
     - allowing dashboard item title to be edited 
     - some more style fixes
     - e2es

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious

### CODE REVIEW

**Comment by liyiy** [comment]
There's a lot of the same code between the dashboard description edit and dashboard item description edit, maybe worth going over at a later time to see which parts to consolidate ^_^

**Comment by liyiy** [comment]
confused with prettier failure that doesn't show what the error is in the `Dashboard.scss` file 😦

File: frontend/src/models/dashboardItemsModel.tsx [hunk 1]
```diff
@@ -3,23 +3,34 @@ import { kea } from 'kea'
...
 import { dashboardsModel } from './dashboardsModel'
 import { Link } from 'lib/components/Link'
 import { dashboardItemsModelType } from '~/models/dashboardItemsModelType'
+import { insightLogic } from 'scenes/insights/insightLogic'
 
 export const dashboardItemsModel = kea<dashboardItemsModelType<DashboardItemType>>({
     actions: () => ({
         renameDashboardItem: (item: DashboardItemType) => ({ item }),
         renameDashboardItemSuccess: (item: DashboardItemType) => ({ item }),
+        updateDashboardItem: (id: number, payload: any) => ({ id, ...payload }),
```
Comment (paolodamico) [inline frontend/src/models/dashboardItemsModel.tsx:16]: The payload could be a `Partial<DashboardItemType>` probably

**Review by paolodamico — CHANGES_REQUESTED** [review]
Looking pretty good!! Reviewed this from a UX perspective,

1. I think we should keep the same consistent behavior for editing as we do for dashboards (see below). Clicking to edit instead of the button, close if you click outside, cmd/cntrl + enter to submit.
2. I think dashboard item might be confusing. Maybe dashboard insight?
3. Related to the above the placeholder text should be: "Add a description for this dashboard insight..."
4. The description should not be muted if it's set (show in default text color). Probably worth having it the same font size as the rest (unsure, let's try it out).
5. You should be able to edit the item's description from the dashboard page.

File: frontend/src/models/dashboardItemsModel.tsx [hunk 2]
```diff
@@ -3,23 +3,34 @@ import { kea } from 'kea'
...
             dashboardId,
             move,
         }),
+        setDashboardItemMode: (mode: DashboardItemMode) => ({ mode }),
         duplicateDashboardItemSuccess: (item: DashboardItemType) => ({ item }),
         refreshAllDashboardItems: (filters: Record<string, any>) => filters,
     }),
+    reducers: () => ({
+        dashboardItemMode: [
+            null,
```
Comment (macobo) [inline frontend/src/models/dashboardItemsModel.tsx:28]: You probably want to set an explicit type here: `null as DashboardItemMode | null`. In general we should avoid implicit any's in typescript.

File: frontend/src/models/dashboardItemsModel.tsx [hunk 3]
```diff
@@ -34,6 +45,15 @@ export const dashboardItemsModel = kea<dashboardItemsModelType<DashboardItemType
...
             })
         },
+        updateDashboardItem: async ({ id, ...payload }) => {
+            if (!Object.entries(payload).length) {
+                return
+            }
+            const response = await api.update(`api/dashboard_item/${id}`, payload)
+            actions.setDashboardItemMode(null)
+            insightLogic.actions.setDashboardItem(response)
+            return response
```
Comment (macobo) [inline frontend/src/models/dashboardItemsModel.tsx:55]: Q: is the return needed here? Why?

Comment (liyiy) [reply frontend/src/models/dashboardItemsModel.tsx:55]: nope! Thought I would need it but I didn't 😃 I shall remove it

File: frontend/src/models/dashboardItemsModel.tsx (hunk 1 above)
Comment (liyiy) [reply frontend/src/models/dashboardItemsModel.tsx:16]: I'm getting `✖ flake8 --select=E9,F63,F7,F82 failed without output (ENOENT)` . errors when I try to merge this branch with master, how do I resolve this?

Comment (paolodamico) [reply frontend/src/models/dashboardItemsModel.tsx:16]: Hmmm not sure what's going on there, but I fix the prettier error and merged master. Sometimes when you try to commit merge commits, husky will fail for some reason, you can try finishing the commit with `--no-verify`.

**Comment by liyiy** [comment]
@macobo I addressed most of the concerns on @paolodamico's comment but in favor of keeping this PR from spiraling, I will address everything else in the second PR of this (editing dashboard insight description on the dashboards page)

**Pull request closed without merging**